File: src/browser_cli/drivers/_extension/input_actions.py
from __future__ import annotations


class ExtensionDriverInputActionsMixin:
    async def type_text(self, page_id: str, text: str, *, submit: bool = False) -> dict[str, object]:
        payload = await self._page_command(page_id, "type", {"text": text, "submit": submit})
        return {"page_id": page_id, **payload}

    async def press_key(self, page_id: str, key: str) -> dict[str, object]:
        payload = await self._page_command(page_id, "press", {"key": key})
        return {"page_id": page_id, **payload}

    async def key_down(self, page_id: str, key: str) -> dict[str, object]:
        payload = await self._page_command(page_id, "key-down", {"key": key})
        return {"page_id": page_id, **payload}

    async def key_up(self, page_id: str, key: str) -> dict[str, object]:
        payload = await self._page_command(page_id, "key-up", {"key": key})
        return {"page_id": page_id, **payload}

    async def wheel(self, page_id: str, *, dx: int = 0, dy: int = 700) -> dict[str, object]:
        payload = await self._page_command(page_id, "scroll", {"dx": dx, "dy": dy})
        return {"page_id": page_id, **payload}

    async def mouse_move(self, page_id: str, *, x: int, y: int) -> dict[str, object]:
        payload = await self._page_command(page_id, "mouse-move", {"x": x, "y": y})
        return {"page_id": page_id, **payload}

    async def mouse_click(
        self,
        page_id: str,
        *,
        x: int,
        y: int,
        button: str = "left",
        count: int = 1,
    ) -> dict[str, object]:
        payload = await self._page_command(
            page_id,
            "mouse-click",
            {"x": x, "y": y, "button": button, "count": count},
        )
        return {"page_id": page_id, **payload}

    async def mouse_drag(self, page_id: str, *, x1: int, y1: int, x2: int, y2: int) -> dict[str, object]:
        payload = await self._page_command(
            page_id,
            "mouse-drag",
            {"x1": x1, "y1": y1, "x2": x2, "y2": y2},
        )
        return {"page_id": page_id, **payload}

    async def mouse_down(self, page_id: str, *, button: str = "left") -> dict[str, object]:
        payload = await self._page_command(page_id, "mouse-down", {"button": button})
        return {"page_id": page_id, **payload}

    async def mouse_up(self, page_id: str, *, button: str = "left") -> dict[str, object]:
        payload = await self._page_command(page_id, "mouse-up", {"button": button})
        return {"page_id": page_id, **payload}

File: browser-cli-extension/src/background/input_actions.js
import {
  keyActionJs,
  mouseActionJs,
  scrollPageJs,
  typeTextJs,
} from '../page_runtime.js';

export function createInputHandlers(context) {
  return {
    async type(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(
        payload.tab_id,
        typeTextJs(payload.text || '', { submit: !!payload.submit }),
      );
    },
    async press(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, keyActionJs('press', payload.key || ''));
    },
    async 'key-down'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, keyActionJs('key-down', payload.key || ''));
    },
    async 'key-up'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, keyActionJs('key-up', payload.key || ''));
    },
    async scroll(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, scrollPageJs(payload.dx || 0, payload.dy || 700));
    },
    async 'mouse-move'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, mouseActionJs('mouse-move', payload));
    },
    async 'mouse-click'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, mouseActionJs('mouse-click', payload));
    },
    async 'mouse-drag'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, mouseActionJs('mouse-drag', payload));
    },
    async 'mouse-down'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, mouseActionJs('mouse-down', payload));
    },
    async 'mouse-up'(payload) {
      await context.getManagedTab(payload.tab_id);
      return await context.executeExpression(payload.tab_id, mouseActionJs('mouse-up', payload));
    },
  };
}
